test/connect.js

```diff
@@ -1,522 +1,450 @@
+    // cmd: 'connect', protocolId: 'MQTT', protocolVersion: 2, clean: true, clientId: 'my-client', keepalive: 0
+    const rawPacket = '10 15 00 04 4D 51 54 54 02 02 00 00 00 09 6D 79 2D 63 6C 69 65 6E 74'
+    rawWrite(s, rawPacket)
+  }
 
-    s.inStream.write({
-      cmd: 'connect',
-      protocolId: 'MQTT',
-      protocolVersion: 2,
-      clean: true,
-      clientId: 'my-client',
-      keepalive: 0
-    })
-    s.outStream.on('data', function (packet) {
-      t.fail('no data sent')
-    })
-    broker.on('connectionError', function (client, err) {
-      t.equal(client.version, null)
-      t.equal(err.message, 'Invalid protocol version')
-      t.equal(broker.connectedClients, 0)
-    })
-  })
+  const checkDisconnect = async () => {
+    const [client, err] = await once(s.broker, 'connectionError')
+    t.assert.equal(client.version, null)
+    t.assert.equal(err.message, 'Invalid protocol version')
+    t.assert.equal(broker.connectedClients, 0)
+  }
+
+  // run parallel
+  await Promise.all([
+    checkNoPacket(t, s),
+    checkDisconnect(),
+    sendPacket()
+  ])
 })
 
-test('reject clients that exceed the keepalive limit', function (t) {
+test('reject clients that exceed the keepalive limit', async (t) => {
   t.plan(3)
 
-  Aedes.createBroker({
-    keepaliveLimit: 100
-  }).then((broker) => {
-    t.teardown(broker.close.bind(broker))
-
-    const s = setup(broker)
-
-    s.inStream.write({
+  const s = await createAndConnect(t, {
+    broker: {
+      keepaliveLimit: 100
+    },
+    connect: {
       cmd: 'connect',
       keepalive: 150
-    })
-    s.outStream.on('data', function (packet) {
-      t.same(packet, {
-        cmd: 'connack',
-        returnCode: 6,
-        length: 2,
-        qos: 0,
-        retain: false,
-        dup: false,
-        topic: null,
-        payload: null,
-        sessionPresent: false
-      }, 'unsuccessful connack, keep alive limit exceeded')
-    })
-    broker.on('connectionError', function (client, err) {
-      t.equal(err.message, 'keep alive limit exceeded')
-      t.equal(broker.connectedClients, 0)
-    })
+    },
+    expectedReturnCode: 6
   })
+
+  const [client, err] = await once(s.broker, 'connectionError')
+  t.assert.ok(client, 'client is defined')
+  t.assert.equal(err.message, 'keep alive limit exceeded')
+  t.assert.equal(s.broker.connectedClients, 0)
 })
 
+// TODO: test fails because Aedes does not reject this
+// once fixed remove {skip:true}
 // Guarded in mqtt-packet
-test('reject clients with no clientId running on MQTT 3.1.0', function (t) {
+test('reject clients with no clientId running on MQTT 3.1.0', { skip: true }, async (t) => {
   t.plan(3)
 
-  Aedes.createBroker().then((broker) => {
-    t.teardown(broker.close.bind(broker))
+  const broker = await Aedes.createBroker()
+  t.after(() => broker.close())
+  const s = setup(broker)
 
-    const s = setup(broker)
+  // s.inStream.write throws an error when trying to write illegal packets
```

Comment:
wondering why this didn't throw before?

Reply:
I think that the s.instream.write error was previously absorbed by the callbacks. The test succeeded because the client was expected to not connect and it did not connect. The reason for not connecting was however not a rejection because of a missing clientid but because the connect is never sent.

Reply:
yeah sometimes it's hard to understand. BTW everything looks good for now 🙏🏼 Well done